tikv/tikv
# raftstore: calculate group id for every peer when it is syn-recover status

### What is changed and how it works?

Issue Number: Close #15784

What's Changed:
In old version, it doesn't re-calculate group-id for the peers created at marjority replication status even it becomes sync-reocver of dr-auto-sync, and the region is always at sync-recover status even the raft-log replication has catched up.
This pr will re-calculate group-id when the current status is sync-reocver, and the region will become sync status when the raft-log replication catches up.

```commit-message
```

### Related changes

- PR to update `pingcap/docs`/`pingcap/docs-cn`:
- Need to cherry-pick to the release branch

### Check List 

Tests 

- Unit test
- Manual test (add detailed scripts or steps below)
       
       //step 1: down 2 voters in primary dc
	opt.DrClusterInfo.DownNodesInDc(ctxCore, "primary", "tikv", 2)
	c := opt.DrClusterInfo
	ctx := ctxCore
	oltppkg.WaitFor(ctx, "after step 1: down primary 2 tikv, workload running", 1*time.Minute)

	//step 2: change cluster to majority
	output, err := oltppkg.BasicClusterExecPdCtl(c.Cluster, "config set replication-mode majority")
	Expect(err).Should(BeNil())
	Expect(strings.Contains(strings.ToLower(output), "succ")).Should(BeTrue())
	fmt.Println(output)
	oltppkg.WaitFor(ctx, "after step2:change to majority,  workload running", 1*time.Minute)

	//step 3: update placement rule to change from voter to learner for the two down tikv
	c.PdURL = Try(c.Cluster.ServiceURL(resource.PDAddr)).(*url.URL)
	err = drutils.SetPlacementRule(ctx, c.Cluster, c.TiupNode,
		drconfig.PlacementRuleDowngrade)
	Expect(err).Should(BeNil())
	oltppkg.WaitFor(ctx, "after step3: update placement rule, change voter to learner,  workload running", 1*time.Minute)

	//step 4 : clean chaos
	c.CleanChaos()
	oltppkg.WaitFor(ctx, "step 4: clean chaos, waiting for workload running", 1*time.Minute)

	//step 5: set back to dr-auto-sync
	output, err = oltppkg.BasicClusterExecPdCtl(c.Cluster, "config set replication-mode dr-auto-sync")
	Expect(err).Should(BeNil())
	Expect(strings.Contains(strings.ToLower(output), "succ")).Should(BeTrue())
	fmt.Println(output)
	oltppkg.WaitFor(ctx, "step5: set back to dr-autosync workload running", 1*time.Minute)

	//step 6: set back to dr-auto-sync placement rule
	err = drutils.SetPlacementRule(ctx, c.Cluster, c.TiupNode,
		drconfig.PlacementRuleConfigPrimary)
	Expect(err).Should(BeNil())
	drutils.WaitForReplicateStatus(ctx, c.Cluster, 10*time.Minute, 1*time.Minute, "sync")
	oltppkg.WaitFor(gctx, "step 6: set back to dr-auto-sync placement rule, waiting for workload running", 5*time.Minute)
       

Side effects

- Performance regression
    - Consumes more CPU
    - Consumes more MEM
- Breaking backward compatibility

### Release note 

```release-note
Please add a release note.

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

If you don't think this PR needs a release note then fill it with None.
If this PR will be picked to release branch, then a release note is probably required.
```
